File: apps/ios/BitOS/DesignSystem/AppButton.swift
import SwiftUI

/**
 * The app button (DESIGN_SYSTEM §8 + mockup `btn` family): filled brand /
 * inverted / outlined / danger-tinted, pill by default, fast pressed
 * scale-down, optional leading SF Symbol, loading state that swaps the
 * icon for a spinner, 45% disabled alpha. Heights clear the 44 pt target.
 */
struct AppButton: View {
    enum Variant {
        case primary
        case onSurface
        case ghost
        case danger
    }

    enum Size {
        case sm
        case md
        case lg

        var height: CGFloat {
            switch self {
            case .sm: return 36
            case .md: return 44
            case .lg: return 52
            }
        }

        var font: Font {
            switch self {
            case .sm: return BitOSType.labelMedium
            case .md, .lg: return BitOSType.labelLarge
            }
        }

        var horizontalPadding: CGFloat {
            switch self {
            case .sm: return BitOSTheme.Spacing.base
            case .md: return BitOSTheme.Spacing.lg
            case .lg: return BitOSTheme.Spacing.xl
            }
        }
    }

    let title: String
    var icon: String? = nil
    var variant: Variant = .primary
    var size: Size = .md
    var pill = true
    var enabled = true
    var loading = false
    var action: () -> Void

    private var container: Color {
        switch variant {
        case .primary: return BitOSTheme.accent
        case .onSurface: return BitOSTheme.textPrimary
        case .ghost: return BitOSTheme.surface
        case .danger: return BitOSTheme.error.opacity(0.12)
        }
    }

    private var content: Color {
        switch variant {
        case .primary: return Color(hex: 0x1A1000)
        case .onSurface: return BitOSTheme.background
        case .ghost: return BitOSTheme.textPrimary
        case .danger: return BitOSTheme.errorText
        }
    }

    private var stroke: Color? {
        switch variant {
        case .ghost: return BitOSTheme.border
        case .danger: return BitOSTheme.error.opacity(0.4)
        default: return nil
        }
    }

    var body: some View {
        Button {
            guard enabled, !loading else { return }
            action()
        } label: {
            HStack(spacing: BitOSTheme.Spacing.sm) {
                if loading {
                    ProgressView()
                        .tint(content)
                        .controlSize(.small)
                } else if let icon {
                    Image(systemName: icon)
                        .font(.system(size: size == .sm ? 13 : 15, weight: .semibold))
                }
                Text(title)
                    .font(size.font)
            }
            .frame(maxWidth: .infinity)
            .frame(minHeight: size.height)
            .padding(.horizontal, size.horizontalPadding)
            .foregroundStyle(content)
            .background(
                ZStack {
                    let shape = RoundedRectangle(
                        cornerRadius: pill ? size.height / 2 : BitOSTheme.Radius.md,
                        style: .continuous
                    )
                    shape.fill(container)
                    if let stroke {
                        shape.strokeBorder(stroke, lineWidth: 1)
                    }
                }
            )
            .opacity(enabled || loading ? 1 : 0.45)
            .contentShape(Rectangle())
        }
        .buttonStyle(PressedScaleStyle())
        .disabled(!enabled || loading)
        .accessibilityLabel(Text(title))
    }
}

/// Fast pressed scale-down shared by tappable brand controls (§2.4).
struct PressedScaleStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.97 : 1)
            .animation(BitOSMotion.standard(BitOSMotion.fast), value: configuration.isPressed)
    }
}

/// Compact icon-only action target (48-equivalent square, same variants).
struct AppIconButton: View {
    let systemIcon: String
    let contentDescription: String
    var variant: AppButton.Variant = .ghost
    var enabled = true
    var action: () -> Void

    private var container: Color {
        switch variant {
        case .primary: return BitOSTheme.accent
        case .onSurface: return BitOSTheme.textPrimary
        case .ghost: return BitOSTheme.surface
        case .danger: return BitOSTheme.error.opacity(0.12)
        }
    }

    private var content: Color {
        switch variant {
        case .primary: return Color(hex: 0x1A1000)
        case .onSurface: return BitOSTheme.background
        case .ghost: return BitOSTheme.textSecondary
        case .danger: return BitOSTheme.errorText
        }
    }

    var body: some View {
        Button {
            guard enabled else { return }
            action()
        } label: {
            Image(systemName: systemIcon)
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(content)
                .frame(
                    width: max(BitOSTheme.minTouchTarget, 40),
                    height: max(BitOSTheme.minTouchTarget, 40)
                )
                .background(Circle().fill(container))
                .contentShape(Circle())
        }
        .buttonStyle(PressedScaleStyle())
        .disabled(!enabled)
        .accessibilityLabel(Text(contentDescription))
    }
}

#Preview("App buttons — both modes") {
    VStack(spacing: BitOSTheme.Spacing.base) {
        AppButton(title: "Zap", icon: "bolt.fill", action: {})
        AppButton(title: "Follow", variant: .onSurface, action: {})
        AppButton(title: "Share", variant: .ghost, action: {})
        AppButton(title: "Delete", variant: .danger, action: {})
        AppButton(title: "Publishing…", loading: true, action: {})
        AppButton(title: "Disabled", enabled: false, action: {})
        HStack(spacing: BitOSTheme.Spacing.md) {
            AppIconButton(systemIcon: "xmark", contentDescription: "Close", action: {})
            AppIconButton(systemIcon: "bolt.fill", contentDescription: "Zap", variant: .primary, action: {})
        }
    }
    .padding(BitOSTheme.Spacing.screen)
    .background(BitOSTheme.background)
    .environment(\.colorScheme, .dark)
}

File: apps/android/src/main/kotlin/space/bitos/app/ui/designsystem/AppButton.kt
package space.bitos.app.ui.designsystem

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.PressInteraction
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import space.bitos.app.ui.theme.BitOSRadius
import space.bitos.app.ui.theme.bitOSColors
import space.bitos.core.design.DesignTokens

enum class AppButtonVariant { PRIMARY, ON_SURFACE, GHOST, DANGER }

/**
 * Heights clear the 48 dp touch target at MD+ (§2.6); SM is a compact
 * case whose interactive minimum Material still enforces.
 */
enum class AppButtonSize(val height: Dp, val horizontalPadding: Dp) {
    SM(40.dp, 16.dp),
    MD(48.dp, 20.dp),
    LG(56.dp, 24.dp),
}

/**
 * Press tracking against the raw interaction stream (version-stable:
 * `InteractionSource.interactions` + `PressInteraction` events).
 */
@Composable
private fun rememberPressed(interaction: MutableInteractionSource): Boolean {
    val pressed = remember { mutableStateOf(false) }
    LaunchedEffect(interaction) {
        interaction.interactions.collect { event ->
            when (event) {
                is PressInteraction.Press -> pressed.value = true
                is PressInteraction.Release,
                is PressInteraction.Cancel -> pressed.value = false
            }
        }
    }
    return pressed.value
}

/**
 * The app button (DESIGN_SYSTEM §8 + mockup `btn` family): filled /
 * inverted / outlined / danger-tinted, pill by default, pressed
 * scale-down (fast, standard easing), optional leading icon, loading
 * state that swaps the icon for a spinner, and 45% disabled alpha.
 */
@Composable
fun AppButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: AppButtonVariant = AppButtonVariant.PRIMARY,
    size: AppButtonSize = AppButtonSize.MD,
    pill: Boolean = true,
    enabled: Boolean = true,
    loading: Boolean = false,
    leadingIcon: @Composable (() -> Unit)? = null,
) {
    val colors = bitOSColors()
    val scheme = MaterialTheme.colorScheme
    val (container, content, borderColor) = when (variant) {
        AppButtonVariant.PRIMARY -> Triple(colors.primary, colors.onPrimary, Color.Transparent)
        AppButtonVariant.ON_SURFACE -> Triple(colors.textPrimary, colors.background, Color.Transparent)
        AppButtonVariant.GHOST -> Triple(colors.surface, colors.textPrimary, colors.border)
        AppButtonVariant.DANGER -> Triple(
            colors.error.copy(alpha = 0.12f),
            colors.errorText,
            colors.error.copy(alpha = 0.4f),
        )
    }
    val shape = if (pill) CircleShape else androidx.compose.foundation.shape.RoundedCornerShape(BitOSRadius.md)
    val interaction = remember { MutableInteractionSource() }
    val pressed = rememberPressed(interaction)
    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.97f else 1f,
        animationSpec = BitOSMotion.standard(BitOSMotion.fast),
        label = "pressScale",
    )
    val interactive = enabled && !loading

    Row(
        modifier = modifier
            .heightIn(min = size.height)
            .scale(scale)
            .clip(shape)
            .background(container, shape)
            .then(if (borderColor != Color.Transparent) Modifier.border(1.dp, borderColor, shape) else Modifier)
            .alpha(if (interactive) 1f else 0.45f)
            .clickable(interactionSource = interaction, indication = null, enabled = interactive, onClick = onClick)
            .padding(horizontal = size.horizontalPadding, vertical = 10.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (loading) {
            CircularProgressIndicator(
                modifier = Modifier.size(if (size == AppButtonSize.SM) 14.dp else 18.dp),
                strokeWidth = 2.dp,
                color = content,
            )
        } else {
            leadingIcon?.invoke()
        }
        Text(
            text = text,
            style = if (size == AppButtonSize.SM) {
                MaterialTheme.typography.labelMedium
            } else {
                MaterialTheme.typography.labelLarge
            },
            color = content,
        )
    }
}

/** Compact icon-only action target: 48 dp square, same variants. */
@Composable
fun AppIconButton(
    onClick: () -> Unit,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    variant: AppButtonVariant = AppButtonVariant.GHOST,
    enabled: Boolean = true,
    icon: @Composable () -> Unit,
) {
    val colors = bitOSColors()
    val container = when (variant) {
        AppButtonVariant.PRIMARY -> colors.primary
        AppButtonVariant.ON_SURFACE -> colors.textPrimary
        AppButtonVariant.GHOST -> colors.surface
        AppButtonVariant.DANGER -> colors.error.copy(alpha = 0.12f)
    }
    val interaction = remember { MutableInteractionSource() }
    val pressed = rememberPressed(interaction)
    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.97f else 1f,
        animationSpec = BitOSMotion.standard(BitOSMotion.fast),
        label = "pressScale",
    )
    Box(
        modifier = modifier
            .size(DesignTokens.MIN_TOUCH_TARGET_DP.dp)
            .scale(scale)
            .clip(CircleShape)
            .background(container, CircleShape)
            .alpha(if (enabled) 1f else 0.45f)
            .clickable(interactionSource = interaction, indication = null, enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) { icon() }
}
